runtime/src/bank/builtins/core_bpf_migration/bpf_upgradeable.rs

```diff
@@ -0,0 +1,373 @@
+        if self.program_data_account.owner != BPF_LOADER_UPGRADEABLE_ID {
+            return Err(CoreBpfMigrationError::IncorrectOwner(
+                self.program_data_address,
+            ));
+        }
+
+        // The program data account should have the correct state.
+        let programdata_data_offset = UpgradeableLoaderState::size_of_programdata_metadata();
+        if self.program_data_account.data.len() < programdata_data_offset {
+            return Err(CoreBpfMigrationError::InvalidProgramDataAccount(
+                self.program_data_address,
+            ));
+        }
+        // Length checked in previous block.
+        match bincode::deserialize::<UpgradeableLoaderState>(
+            &self.program_data_account.data[..programdata_data_offset],
+        ) {
+            Ok(UpgradeableLoaderState::ProgramData { .. }) => Ok(()),
+            _ => Err(CoreBpfMigrationError::InvalidProgramDataAccount(
+                self.program_data_address,
+            )),
+        }
+    }
+
+    /// Create a new migration configuration for a BPF Upgradeable source
+    /// program.
+    pub(crate) fn new_checked(
+        bank: &Bank,
+        program_id: &Pubkey,
+    ) -> Result<Self, CoreBpfMigrationError> {
+        let program_address = *program_id;
+        // The program account should exist.
+        let program_account: Account = bank
+            .get_account_with_fixed_root(&program_address)
+            .ok_or(CoreBpfMigrationError::AccountNotFound(program_address))?
+            .into();
+
+        // The program data account should exist.
+        let program_data_address = get_program_data_address(&program_address);
+        let program_data_account: Account = bank
+            .get_account_with_fixed_root(&program_data_address)
+            .ok_or(CoreBpfMigrationError::ProgramHasNoDataAccount(
+                program_address,
+            ))?
+            .into();
+
+        // The total data size is the size of the program account's data plus
+        // the size of the program data account's data.
+        let total_data_size = program_account
+            .data
+            .len()
+            .checked_add(program_data_account.data.len())
+            .ok_or(CoreBpfMigrationError::ArithmeticOverflow)?;
+
+        let config = Self {
+            program_address,
+            program_account,
+            program_data_address,
+            program_data_account,
+            total_data_size,
+        };
+
+        config.check_program_account()?;
+        config.check_program_data_account()?;
+
+        Ok(config)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        crate::bank::tests::create_simple_test_bank,
+        solana_sdk::{
+            account::AccountSharedData, bpf_loader_upgradeable::ID as BPF_LOADER_UPGRADEABLE_ID,
+        },
+    };
+
+    fn store_account<T: serde::Serialize>(
+        bank: &Bank,
+        address: &Pubkey,
+        data: (&T, Option<&[u8]>),
```

Comment:
nit: any reason for passing the data as a tuple over two separate parameters?

Reply:
Mostly because it made life easier when serializing the program data account's initial state before appending the program ELF. There's one test where I intentionally serialize the wrong header. But I can look into simplifying for readability.

Reply:
Ahh I misread this. Your concern is with the tuple specifically, not the separation. Yeah OK, I can change that.